### PR TITLE
feat: Add Auth Context enabled Firestore triggers

### PR DESCRIPTION
**Pull request by exaby73**

### CODE REVIEW

File: src/firebase_functions/firestore_fn.py [hunk 1]
```diff
@@ -146,34 +172,49 @@ def _firestore_endpoint_handler(
...
+    if event_type.endswith(".withAuthContext"):
+        database_event_with_auth_context = EventWithAuthContext(
+            **common_event_kwargs,
+            auth_type=event_auth_type,
+            auth_id=event_auth_id)
+        func(database_event_with_auth_context)
+    else:
+        database_event = Event(**common_event_kwargs)
+        # mypy cannot infer that the event type is correct, hence the cast
+        _typing.cast(_C1 | _C2, func)(database_event)
```
Comment (pr-Mais) [inline src/firebase_functions/firestore_fn.py:217]: Can probably keep this as an `Event` type, then if it's auth context event use `__dict__ ` to append it to `EventWithAuthContext`, something like:
```suggestion
   database_event = Event(
        project=event_project,
        namespace=event_namespace,
        database=event_database,
        location=event_location,
        document=event_document,
        specversion=event_attributes["specversion"],
        id=event_attributes["id"],
        source=event_attributes["source"],
        type=event_attributes["type"],
        time=event_time,
        data=firestore_event_data,
        subject=event_attributes["subject"],
        params=params,
    )

    if event_type.endswith(".withAuthContext"):
        database_event_with_auth_context = EventWithAuthContext(
            **vars(database_event),
            auth_type=event_auth_type,
            auth_id=event_auth_id)
        func(database_event_with_auth_context)
    else:
        func(database_event)
```

**Review by pr-Mais — COMMENTED** [review]
LGTM, we just need some unit tests if feasible.

File: src/firebase_functions/firestore_fn.py [hunk 2]
```diff
@@ -82,9 +87,26 @@ class Event(_core.CloudEvent[_core.T]):
 _C1 = _typing.Callable[[_E1], None]
 _C2 = _typing.Callable[[_E2], None]
 
+AuthType = _typing.Literal["service_account", "api_key", "system",
+                           "unauthenticated", "unknown"]
+
+
+@_dataclass.dataclass(frozen=True)
+class EventWithAuthContext(Event[_core.T]):
```
Comment (blidd-google) [inline src/firebase_functions/firestore_fn.py:95]: FYI we ended up deciding to call the separate event interface FirestoreAuthEvent in the Node SDK. I guess the equivalent here would be AuthEvent?

**Review by blidd-google — COMMENTED** [review]
LGTM except nit re: naming

**Review by pr-Mais — APPROVED** [review]

**Review by blidd-google — APPROVED** [review]

**Pull request merged**